[Snackbar] Display stacked Snackbars

ZeeshanTamboli opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->
- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

Resolves mui/material-ui#1824
Resolves mui/material-ui#18098

This is an opt-in feature.

This work is a replacement of [notistack](https://github.com/iamhosseindhv/notistack).

Demo preview: https://deploy-preview-31991--material-ui.netlify.app/material-ui/react-snackbar/#experimental-api

~Initial implementation and idea could be reviewed.~

~Should I move it to MUI lab? If yes, how to reuse the `Snackbar` component styles into `SnackbarsProvider` styles without duplicating code by exporting from `mui/material` to `mui/lab` but not providing it to user in `mui/material`?~

## TODO
- [x] Implement other props in `SnackbarsProvider`. 
- [x] Add tests
- [x] Improve docs
- [x] Fix typescript types (Remove `@ts-expect-error` and fix types)
- [x] Move to lab (files  + docs) (https://github.com/mui/material-ui/pull/31991#issuecomment-1087294766)

----

Edit: Ready for review!

mnajdova commented:
> Should I move it to MUI lab?

Yes.

> If yes, how to reuse the Snackbar component styles into SnackbarsProvider styles without duplicating code by exporting from mui/material to mui/lab but not providing it to user in mui/material?

I wouldn't care for this just yet, let's duplicate.

ZeeshanTamboli commented:
@mui/core I will be waiting for a review before I continue further to add tests.

ZeeshanTamboli commented on docs/translations/api-docs/snackbars-provider/snackbars-provider-pt.json:
```diff
@@ -0,0 +1,16 @@
+{
+  "componentDescription": "",
+  "propDescriptions": {
+    "action": "The action to display. It renders after the message, at the end of the snackbar.",
```
> Seeing the props it is not clear which ones are for the SnackbarProvider itself, and which are going to be applied on the snackbars generated

All the props of the `SnackbarsProvider` are available in the snackbars generation method (`showSnackbar`) except `limit` which is only available in `SnackbarsProvider`. Likewise, the `message` option to show the message of the snackbar is only available in `showSnackbar` and not in `SnackbarsProvider`. 
For the mutual properties in `SnackbarsProvider` and `showSnackbar` options , the `showSnackbar` options for individual snackbar takes more priority over props in `SnackbarsProvider` (if props are present in both). 
I will be documenting them clearly.

> Could we maybe threat the snackbars as a "slot" inside and add the props there? For example:
componentsProps.snackbar = {/* all props related to the snackbar */}

I think `componentsProps` makes more sense only when there are multiple slots/components inside the component on which the prop is passed. `SnackbarsProvider` is for snackbars only .

ZeeshanTamboli commented on docs/data/material/components/snackbars/snackbars.md:
```diff
@@ -18,7 +18,7 @@ They may contain a text action, but no icons. You can use them to display notifi
 {{"component": "modules/components/ComponentLinkHeader.js"}}
 
-**Frequency**: Only one snackbar may be displayed at a time.
+**Frequency**: Only one snackbar may be displayed at a time. To display more than one snackbars, go to [multiple snackbars](/components/multiple-snackbars/) section.
```
Hmm, I see this as the correct URL for redirecting on other components in the markdown pages. But this path does not work (https://deploy-preview-31991--material-ui.netlify.app/material-ui/react-multiple-snackbars). For other components it works.
As per my investigation this path is for `Material-UI -> MUI/Core` product page. Is this something related to migration being pending? I have kept it as `/components/mutiple-snackbars/` for now.

ZeeshanTamboli replied:
Fixed now.

ZeeshanTamboli commented on packages/mui-lab/src/SnackbarsProvider/SnackbarsProvider.tsx:
```diff
@@ -0,0 +1,280 @@
+import SnackbarsContext, { ShowSnackbarProps, SnackbarContentType } from './SnackbarsContext';
+import SnackbarsContainer from './SnackbarsContainer';
+
+const randomId = () => `mui-${Math.round(Math.random() * 1e5)}`;
```
Do we have a id generation `util` method that I can reuse here? I tried `useId`, it does not work here.

AmirMehr commented:
I have developed a simple `Snackbar` repo, you can add it to the MUI for showing multiple `Snackbar`. If you do so, no need to install `notistack`. Here is the repo: [https://github.com/AmirMehr/MaterialNotify](https://github.com/AmirMehr/MaterialNotify)

siriwatknp commented:
Thanks for kicking off this effort. I'd like to have some clarification on these questions:

**Roll out strategy**

My understanding is that this work is a replacement of [notistack](https://github.com/iamhosseindhv/notistack) but it is still blur to me on other aspects apart from the implementation like the goal, plan, migration. I think these topics are quite as important as the implementation so we should be able to answer them before rolling it out to the public (even in the lab). Can you share what's on your mind?

**UX**

It is too stiff when the noti leave the screen. If you compare with notistack, it is more smooth.

https://user-images.githubusercontent.com/18292247/165260826-be84f384-b40b-4d8d-bb93-ca8d35e79765.mov

ZeeshanTamboli commented:
@siriwatknp 

> Thanks for kicking off this effort. I'd like to have some clarification on these questions:
> 
> **Roll out strategy**
>
> My understanding is that this work is a replacement of [notistack](https://github.com/iamhosseindhv/notistack) but it is still blur to me on other aspects apart from the implementation like the goal, plan, migration. I think these topics are quite as important as the implementation so we should be able to answer them before rolling it out to the public (even in the lab). Can you share what's on your mind?

Yes, the APIs are mostly similar to notistack which is built on top of MUI Snackbar component. We should mention this in the doc like: `Inspired by notistack`. 

Although it is discouraged to have multiple snackbars as mentioned in material design specification, 1. https://material.io/components/snackbars#usage, 2. https://github.com/mui/material-ui/issues/1824#issuecomment-278839071, many people  still need to display multiple snackbars in their applications as per the discussion in #1824 and the imperative API style as discussed in #18098 which this PR tries to solve.

I think notistack will be focusing more on notifications without Material UI as per https://github.com/iamhosseindhv/notistack/issues/184 and https://github.com/iamhosseindhv/notistack/issues/421 (although it's in alpha still). Can't be sure when this will be released.

There will be no need to install a separate package outside of MUI if MUI itself supports it (See [remove the dependency on notistack](https://github.com/iamhosseindhv/notistack/issues/184#issuecomment-563189013)). 

Many people still use notistack for the material design but we can focus more in terms of:
1. Customization with the APIs MUI uses for customizing components.
2. Supporting `SnackbarProvider` in global theme.
3. Perhaps extend it to base apart from unstyled single `Snackbar` component? 
4. Providing continuous maintenance.

Migration is not needed anywhere as this an opt-in feature using the `Snackbar` component.

> **UX**
>
> It is too stiff when the noti leave the screen. If you compare with notistack, it is more smooth. 

I can take a look at this if we agree to support it in MUI even though there is notistack.

mnajdova commented:
I strongly believe this is a feature we are leaking to have in house, for example [Chakra UI](https://chakra-ui.com/docs/components/feedback/toas),  [Ant design](https://ant.design/components/notification/) and [Mantine](https://mantine.dev/others/notifications/) all support it, so I would propose continuing the effort @ZeeshanTamboli (some inspiration can probably be taken from these libraries in terms of implementation).

I agree with @siriwatknp that we should have some migration story in mind from notistack to this new API, simply because we were the ones suggesting developers to use notistack in our documentation.

ZeeshanTamboli commented:
> UX
>
> It is too stiff when the noti leave the screen. If you compare with notistack, it is more smooth.

@siriwatknp Did you compare notistack with https://deploy-preview-31991--material-ui.netlify.app/material-ui/react-multiple-snackbars/#maximum-snackbars? Our demo uses the `Grow` transition which is the default transition of Snackbar. What `notistack` has is a Slide transition towards the right direction. We have a transition section below.

> I strongly believe this is a feature we are leaking to have in house, for example [Chakra UI](https://chakra-ui.com/docs/components/feedback/toas), [Ant design](https://ant.design/components/notification/) and [Mantine](https://mantine.dev/others/notifications/) all support it, so I would propose continuing the effort @ZeeshanTamboli (some inspiration can probably be taken from these libraries in terms of implementation).

@mnajdova Implementation can be reviewed further. I have implemented almost all of the cases. Anything which is important in the primary version can be added. For other cases, I think it can be added later one at a time after more feedback from users. Do check the implementation and the preview docs and let me know what you think.

> I agree with @siriwatknp that we should have some migration story in mind from notistack to this new API, simply because we were the ones suggesting developers to use notistack in our documentation.

We can have a migration guide section in the docs. The installation steps involved, the API names from notistack to this, etc.

mnajdova commented on docs/data/material/components/multiple-snackbars/multiple-snackbars.md:
```diff
@@ -0,0 +1,89 @@
+waiAria: https://www.w3.org/TR/wai-aria-1.1/#alert
+---
+
+# Multiple Snackbars
```
@samuelsycamore could you please review the content of this page?

mnajdova replied:
Also, what should we do with:
1. https://mui.com/material-ui/react-snackbar/#consecutive-snackbars
2. https://deploy-preview-31991--material-ui.netlify.app/material-ui/react-snackbar/#notistack

danilo-leal replied:
I've done a bit of writing revision there but would love Sam's input as well regardless!

I've got a bit confused by the name in relation to what's being introduced. As far as I understood, there's a new `SnackbarsProvider` component and a `useSnackbars` hook. First question would be: wouldn't make more sense for the hook to be in Base? Second is: instead of placing the `SnackbarsProvider` in the lab, shouldn't we add it as a supporting component for the Snackbar?

That would allow having all the content inside the Snackbar, which seems to make a lot more sense to me. We'd avoid content and demo repetition while also connecting the narrative that notistack used to be what enabled multiple snackbars but now we have a "native" solution for that, "deprecating" the need to depend on notistack for that.

ZeeshanTamboli replied:
`useSnackbars` is not a low level hook for unstyled Snackbar to be it in Base. It's a hook available for the applications where it can be used within their components to trigger a snackbar inside a `SnackbarsProvider` context.

While I agree that I will be easier to manage if it is moved within core's Snackbar for all the reasons you mentioned, it is supposed to be in lab since this is first version and we can would get more feedback from users regarding bugs, design, API design which is what lab is used for initially. 
Once it's robust, we move it to core which will make things much easier then (regarding repetitive demos etc.)

For now we can still show there is a replacement of notistack available in lab by having a link to this page. I have also added a link to multiple snackbars from Core's snackbar [here](https://deploy-preview-31991--material-ui.netlify.app/material-ui/react-snackbar/#main-content) in the `Frequency` section.  

What do you think?

danilo-leal replied:
Right, that makes sense. I changed the word slightly from _multiple_ to _stacked_ snackbars, which sounds clearer to me. Additionally, edited the documentation to showcase that notistack and now this component are both alternatives to stack snackbars, trying to make it clear that the latter is our native take on it.

mnajdova reviewed (commented):
Looks great in general. I left some comments here and there. The next step would be polishing the design (transitions) and docs. cc @danilo-leal @samuelsycamore 

Great job @ZeeshanTamboli 👌

danilo-leal commented on docs/data/material/components/multiple-snackbars/multiple-snackbars.md:
```diff
@@ -0,0 +1,86 @@
+
+## Props priority
+
+For the mutual properties between `showSnackbar` options and `SnackbarsProvider` props, the options in `showSnackbar` method of `useSnackbars` hook for individual snackbars take more preference over props passed in `SnackbarsProvider` if both are supplied.
```
I've had a hard time trying to understand this paragraph and I think I haven't yet 😅 Is there a way we could try to simplify it?

ZeeshanTamboli replied:
I agree it is confusing, how about this:
```suggestion
In the event, when there are same props supplied for both `showSnackbar` and `SnackbarsProvider`, the options in `showSnackbar` method will gain priority over props in `SnackbarsProvider`. 
```

ZeeshanTamboli commented on docs/data/material/components/multiple-snackbars/GlobalContentSnackbars.js:
```diff
@@ -0,0 +1,47 @@
+      ref={snackbarRef}
+      autoHideDuration={4000}
+      anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
+      content={(key) => (
```
It's shown in the `Customization` section where the `content` prop can be used to modify each individual snackbar.

danilo-leal reviewed (commented):
This is looking great so far, amazing work! 
I'd just wait for @samuelsycamore perspective on it as well before merging =)

oliviertassinari commented on docs/data/material/components/snackbars/snackbars.md:
```diff
@@ -18,8 +18,6 @@ They may contain a text action, but no icons. You can use them to display notifi
 
 {{"component": "modules/components/ComponentLinkHeader.js"}}
 
-**Frequency**: Only one snackbar may be displayed at a time.
```
This is still the recommendation of Material Design, proof: https://material.io/components/snackbars#usage. We could make it clearer, but I think that the message should still be present.

ZeeshanTamboli replied:
@danilo-leal Can you take a look?

oliviertassinari commented on docs/data/material/components/snackbars/snackbars.md:
```diff
@@ -40,6 +38,26 @@ You can control the position of the snackbar by specifying the `anchorOrigin` pr
+
+We have been developing a native component to support this use case and [we've made it available in the lab](/material-ui/react-stacked-snackbars/) to gather feedback for its first versions and general API design.
+
+### notistack
+
+![stars](https://img.shields.io/github/stars/iamhosseindhv/notistack.svg?style=social&label=Stars)
+![npm downloads](https://img.shields.io/npm/dm/notistack.svg)
+
+This example demonstrates how to use [notistack](https://github.com/iamhosseindhv/notistack).
+notistack has an **imperative API** that makes it easy to display snackbars, without having to handle their open/close state.
+It also enables you to **stack** them on top of one another (although this is discouraged by the Material Design guidelines).
+
+{{"demo": "IntegrationNotistack.js", "defaultCodeOpen": false}}
```
It sounds like we can replace all mentions of notistack in the docs now

```suggestion
```

ZeeshanTamboli replied:
@danilo-leal Can you take a look?

danilo-leal replied:
Addressed in https://github.com/mui/material-ui/pull/31991/commits/7a351f66c7efecb43be56c34a782349da7494594  👍

oliviertassinari commented on docs/data/material/components/stacked-snackbars/GlobalContentSnackbars.tsx:
```diff
@@ -0,0 +1,50 @@
+
+  return (
+    <React.Fragment>
+      <Button onClick={() => snackbars.showSnackbar()}>
```
Would a shorter API work as well?

```suggestion
      <Button onClick={() => snackbars.show()}>
```

What's the need to have "snackbar" in the method?

ZeeshanTamboli replied:
Yes we can make it shorter. 

Some benchmarks:
1. notistack - `const { enqueueSnackbar, closeSnackbar } = useSnackbar();`
2. mantine - `import { showNotification, updateNotification } from '@mantine/notifications';`

So for us, `snackbars.show()`, `snackbars.close()` makes sense after having `const snackbars = useSnackbars()`. I will do the changes.

ZeeshanTamboli replied:
Renamed!

oliviertassinari commented on packages/mui-lab/src/SnackbarsProvider/SnackbarsProvider.tsx:
```diff
@@ -0,0 +1,279 @@
+
+export interface SnackbarsProviderProps
+  extends Omit<SnackbarProps, 'children' | 'classes' | 'key' | 'message' | 'onClose' | 'open'> {
+  /** The maximum number of snackbars to display at a time.
```
```suggestion
  /**
   * The maximum number of snackbars to display at a time.
```

oliviertassinari commented on docs/src/modules/utils/replaceUrl.ts:
```diff
@@ -39,7 +39,7 @@ export const replaceComponentLinks = (url: string) => {
   } else {
     url = url.replace(/\/components\/(.*)/, '/material-ui/react-$1');
-    if (!url.match(/\/react-(tabs|breadcrumbs)/)) {
+    if (!url.match(/\/react-(tabs|breadcrumbs|stacked-snackbars)/)) {
```
This should not be necessary anymore. I believe that all of this logic is legacy, can be deleted.

```suggestion
    if (!url.match(/\/react-(tabs|breadcrumbs)/)) {
```

ZeeshanTamboli replied:
API build script (`yarn docs:api`) is traversing this part of the code. As far as I looked, it is replacing the docs components filename path urls from plural to singular. 
Example:
`/components/checkboxes/` -> `/material-ui/react-checkbox/` (Note the `checkboxes` (plural) gets converted to singular `checkbox`). The components urls should be named singularly. Not `Checkboxes`, `Buttons`.

Whereas, this condition prevents `Tabs`, `Breadcumbs` and `Stacked snackbars` to be changed to singular word links. These should be plural.  
Example:
`/components/tabs/` -> `/material-ui/react-tabs/`
`/components/breadcumbs/` -> `/material-ui/react-breadcumbs/`

I added a test assertion for it in 481305c264f253255276006ef40a2a7f0e6102bd.

siriwatknp replied:
Yes, I'm on it.

oliviertassinari commented on packages/mui-lab/src/SnackbarsProvider/SnackbarsProvider.tsx:
```diff
@@ -0,0 +1,279 @@
+export interface SnackbarsProviderProps
+  extends Omit<SnackbarProps, 'children' | 'classes' | 'key' | 'message' | 'onClose' | 'open'> {
+  /** The maximum number of snackbars to display at a time.
+   * @default 3
```
Why have a limit by default? All the libraries I could benchmark don't have a limit by default

danilo-leal replied:
I think it's reasonable to limit by default, otherwise it'd be a non-functional UX pattern? I don't feel like stacking a bunch of snackbars would be useful, it seems just bloated noise in the UI. 

Though I see that we could simply just recommend that instead of actually limiting from a component-level. 🤔

oliviertassinari replied:
Alright, I simply found it curious that all the alternatives don't have a limit but we do. I think that time will tell us if it works or not. We can simply stay attentive to the UX feedback we will get on the GitHub issues 👍

ZeeshanTamboli replied:
`notistack` has default limit of 3 and `mantine` has 5. 

>  I don't feel like stacking a bunch of snackbars would be useful, it seems just bloated noise in the UI.

This is what came to my mind and thus added a limit after benchmarking.

oliviertassinari commented on docs/data/material/components/stacked-snackbars/MaximumSnackbars.tsx:
```diff
@@ -0,0 +1,55 @@
+  const [snackbarLimit, setSnackbarLimit] = React.useState<number>(2);
+  return (
+    <SnackbarsProvider limit={snackbarLimit}>
+      <MyApp snackbarLimit={snackbarLimit} setSnackbarLimit={setSnackbarLimit} />
```
I think that we can improve this demo. We can remove the -/+ buttons

<img width="155" alt="Screenshot 2022-05-23 at 01 29 39" src="https://user-images.githubusercontent.com/3165635/169720687-fe9c7316-8312-400d-9650-5b22eed893ed.png">
 
It would make it much easier to find the interesting part of this demo. Ideally, we would have this:

Then, for developers who want to try different values: #32107 solves the problem. I would argue the DX would be much better. See

![May-23-2022 01-36-38](https://user-images.githubusercontent.com/3165635/169720912-3921eae6-0176-46ff-babd-b8ff8d0e8f34.gif)

oliviertassinari replied:
Do we need to optimize for non-devs? I mean, would a designer want to play around with this demo? I think that we should have the backend devs as our primary target when we write the documentation. For them, what would yield the best DX?

> I think I'd wait for that PR to be merged in order to remove the interactive buttons here.

Opening a new GitHub issue to do clean up sounds great; I think that there will be a good number of interactive demos in our docs that we can remove (improve) thanks to https://github.com/mui/material-ui/pull/32107, this is exciting! :)

oliviertassinari commented on docs/data/material/components/snackbars/snackbars.md:
```diff
@@ -18,8 +18,6 @@ They may contain a text action, but no icons. You can use them to display notifi
 
-**Frequency**: Only one snackbar may be displayed at a time.
-
 ## Simple snackbars
```
Please don't use "simple". This notion differs from one individual to the other
```suggestion
## Basic snackbars
```

Now, since this is not the first h2, I assume this is not even what we want to convey as a message, maybe?

```suggestion
## Action buttons
```

ZeeshanTamboli replied:
> Please don't use "simple". This notion differs from one individual to the other

This is the section in the single `<Snackbar />` component docs. I would not like to change it in this PR. 

Also why `Action buttons`? I didn't get it. It's a snackbar.

danilo-leal replied:
Addressed in https://github.com/mui/material-ui/pull/31991/commits/b94cd195cdeee2fbb1e44bb4bb178c380d3e9a76 👍

oliviertassinari commented on packages/mui-material/src/SnackbarContent/SnackbarContent.d.ts:
```diff
@@ -9,7 +9,7 @@ export interface SnackbarContentProps extends StandardProps<PaperProps, 'childre
    * The action to display. It renders after the message, at the end of the snackbar.
    */
-  action?: React.ReactNode;
+  action?: React.ReactNode | ((key: string) => React.ReactNode);
```
I don't follow this change. `<SnackbarContent>` doesn't seem to support this.

```suggestion
  action?: React.ReactNode;
```

ZeeshanTamboli replied:
It's added to support callback type in `action` method. See: 
```jsx
<SnackbarsProvider
  ref={snackbarRef}
  autoHideDuration={4000}
  action={(key) => (
    <React.Fragment>
      <Button
        color="secondary"
        size="small"
        onClick={snackbarRef.current!.close(key)}
      >
        UNDO
      </Button>
      <IconButton
        size="small"
        aria-label="close"
        color="inherit"
        onClick={snackbarRef.current!.close(key)}
      >
        <CloseIcon fontSize="small" />
      </IconButton>
    </React.Fragment>
  )}
>
  <MyApp />
</SnackbarsProvider>;
```

ZeeshanTamboli replied:
Isolated action prop in https://github.com/mui/material-ui/pull/31991/commits/0e7d8978b6d904cfab6a2d27e9a9b2557b9c7b9c.

oliviertassinari commented on packages/mui-lab/src/useSnackbars/useSnackbars.ts:
```diff
@@ -0,0 +1,12 @@
+  const context = React.useContext(SnackbarsContext);
+
+  if (context === undefined) {
+    throw new Error('MUI: useSnackbars must be within a SnackbarsProvider');
```
```suggestion
    throw new Error('MUI: useSnackbars must be within a SnackbarsProvider/');
```

I think that we could improve this error message. We could differentiate what went wrong to how to fix the problem, and for fixing the problem, maybe even link to the docs with a https://mui.com/r/y link.

I also wonder if the lab shouldn't use `MuiError`.

ZeeshanTamboli replied:
Improved error message in https://github.com/mui/material-ui/pull/31991/commits/834bf996837777ae65f99948c41b35f9d0286e5b and https://github.com/mui/material-ui/pull/31991/commits/4ba165f2da2d108dfbdfba2bc16411da4cd2c353

oliviertassinari reviewed (commented):
@ZeeshanTamboli I don't/won't have the time to dive into detail in this PR but for what I could see so far, we are on the right track 👍. 

Out of curiosity, I have found some of the leader in this space: https://npm-stat.com/charts.html?package=react-hot-toast&package=notistack&package=react-toastify&from=2021-05-22&to=2022-05-22. Likely projects we can learn from.

It seems that we miss the @mui/base story for this feature. I would be in favor we start there instead of with Material UI: so we build at the lower level first. The upvotes in https://github.com/iamhosseindhv/notistack/issues/184 suggest that there could be a large audience for this

<img width="100" alt="Screenshot 2022-05-23 at 01 48 33" src="https://user-images.githubusercontent.com/3165635/169721349-d31f7bc3-bacb-4c4d-91e3-135ea6b45f49.png">

samuelsycamore commented on docs/data/material/components/stacked-snackbars/stacked-snackbars.md:
```diff
@@ -0,0 +1,91 @@
+## Usage
+
+To display multiple snackbars, wrap your application with the `SnackbarsProvider` component.
+That's what will enable you to use the `useSnackbars` hook.
```
Replace "enable" with "lets" and shift to present tense.

```suggestion
This lets you implement the `useSnackbars` hook.
```

samuelsycamore commented on docs/data/material/components/stacked-snackbars/stacked-snackbars.md:
```diff
@@ -0,0 +1,91 @@
+To display multiple snackbars, wrap your application with the `SnackbarsProvider` component.
+That's what will enable you to use the `useSnackbars` hook.
+
+If you're using `ThemeProvider`, wrap `SnackbarsProvider` below it, as follows:
```
I think it's more accurate to say it's "nested within" vs "wrapped below."

```suggestion
If you're using `ThemeProvider`, then `SnackbarsProvider` should be nested inside, as shown here:
```

samuelsycamore commented on docs/data/material/components/stacked-snackbars/stacked-snackbars.md:
```diff
@@ -0,0 +1,91 @@
+
+## Maximum snackbars
+
+The `limit` prop can be passed to `SnackbarsProvider` to limit the maximum number of snackbars displayed at the same time.
```
Rephrasing passive voice for active voice.

```suggestion
You can add the `limit` prop to`SnackbarsProvider` to set the maximum number of snackbars that can be displayed simultaneously.
```

samuelsycamore commented on docs/data/material/components/stacked-snackbars/stacked-snackbars.md:
```diff
@@ -0,0 +1,91 @@
+
+## Basic stacked snackbars
+
+The `message` prop is used to show the snackbar message.
```
Rephrasing for active voice and clarity.

```suggestion
The `message` prop defines the snackbar's text.
```

samuelsycamore commented on docs/data/material/components/stacked-snackbars/stacked-snackbars.md:
```diff
@@ -0,0 +1,91 @@
+
+The `message` prop is used to show the snackbar message.
+
+The `action` prop can be used to show content on the right side of the snackbar.
```
Rephrasing for active voice, and to be more specific (based on my understanding of this prop).

```suggestion
You can add the `action` prop to display additional content on the right side of the snackbar (e.g. a call to action such as an **OK** or **Undo** button).
```

samuelsycamore commented on docs/data/material/components/stacked-snackbars/stacked-snackbars.md:
```diff
@@ -0,0 +1,91 @@
+The `message` prop is used to show the snackbar message.
+
+The `action` prop can be used to show content on the right side of the snackbar.
+It can simply be a React Node or a callback function with a key parameter returning a React Node.
```
Cut "simply."

```suggestion
It can be a React Node, or a callback function with a key parameter that returns a React Node.
```

samuelsycamore commented on docs/data/material/components/stacked-snackbars/stacked-snackbars.md:
```diff
@@ -0,0 +1,91 @@
+It can simply be a React Node or a callback function with a key parameter returning a React Node.
+
+Use the `close` method to dismiss a specific snackbar by providing the `key` from the `action` callback argument.
+
```
Let's add a line here to lead into the demo.

```suggestion

The following demo shows how to work with these basic features:

```

samuelsycamore commented on docs/data/material/components/stacked-snackbars/stacked-snackbars.md:
```diff
@@ -0,0 +1,91 @@
+
+## Customization
+
+Use the `content` prop to customize the individual snackbars.
```
Lead into the demo.

```suggestion
Use the `content` prop to customize individual snackbars, as shown in this demo:
```

samuelsycamore replied:
Customize how, exactly? Is this what makes it possible to add custom components (in this case the `<Alert>`) inside of an individual snackbar with this prop? I think we need to be more specific here.

ZeeshanTamboli replied:
Yes, we can add any custom components, like here it's `<Alert />`. How about this:
```suggestion
Use the `content` prop to customize the individual snackbars. You can build your own custom component and render it by passing it to the `content` prop. 
The following demo renders the `Alert` component in the snackbar by passing it to the `content` prop:
```

samuelsycamore commented on docs/data/material/components/stacked-snackbars/stacked-snackbars.md:
```diff
@@ -0,0 +1,91 @@
+
+{{"demo": "CustomizedSnackbars.js", "defaultCodeOpen": false}}
+
+## Positioned snackbars
```
```suggestion
## Snackbar position
```

samuelsycamore commented on docs/data/material/components/stacked-snackbars/stacked-snackbars.md:
```diff
@@ -0,0 +1,91 @@
+
+## Positioned snackbars
+
+You can control the position of an individual snackbar by specifying the `anchorOrigin` prop.
```
Active voice.

```suggestion
Add the 'anchorOrigin' prop to control the position of an individual snackbar.
```

samuelsycamore commented on docs/data/material/components/stacked-snackbars/stacked-snackbars.md:
```diff
@@ -0,0 +1,91 @@
+## Positioned snackbars
+
+You can control the position of an individual snackbar by specifying the `anchorOrigin` prop.
+For even greater convience, pass the desired `anchorOrigin` value to the `SnackbarsProvider` component to set the position of all snackbars used globally.
```
Rephrasing to put the most important part first.

```suggestion
To set the position of all snackbars used globally, pass the desired `anchorOrigin` value to the `SnackbarsProvider` component.
```

samuelsycamore commented on docs/data/material/components/stacked-snackbars/stacked-snackbars.md:
```diff
@@ -0,0 +1,91 @@
+
+You can control the position of an individual snackbar by specifying the `anchorOrigin` prop.
+For even greater convience, pass the desired `anchorOrigin` value to the `SnackbarsProvider` component to set the position of all snackbars used globally.
+
```
Add a line to introduce the demo.

```suggestion

The following demo shows how to control the positions of individual snackbars:

```

samuelsycamore commented on docs/data/material/components/stacked-snackbars/stacked-snackbars.md:
```diff
@@ -0,0 +1,91 @@
+
+## Transition
+
+Similarly to positioning, use the`TransitionComponent` prop in an individual snackbar if you want to control that specific instance transition effect or use it in the `SnackbarsProvider` component to change it globally.
```
Move the most important part to the front and cut the unnecessary reference to the previous section so it stands on its own.

```suggestion
You can use the`TransitionComponent` prop to control the transition effects on individual snackbars, or use it in the `SnackbarsProvider` component for global effects.
```

samuelsycamore commented on docs/data/material/components/stacked-snackbars/stacked-snackbars.md:
```diff
@@ -0,0 +1,91 @@
+## Transition
+
+Similarly to positioning, use the`TransitionComponent` prop in an individual snackbar if you want to control that specific instance transition effect or use it in the `SnackbarsProvider` component to change it globally.
+
```
Introduce the demo.

```suggestion

The following demo shows how to apply unique transition effects to each individual snackbar in an app:

```

samuelsycamore commented on docs/data/material/components/stacked-snackbars/stacked-snackbars.md:
```diff
@@ -0,0 +1,91 @@
+
+## Props priority
+
+When the same props are supplied for both `show` method of `useSnackbars` and `SnackbarsProvider`, the ones added in the `show` method will have priority over those in `SnackbarsProvider`.
```
Rephrasing for clarity.

```suggestion
Props supplied to the `show` method of an individual snackbar take priority over those same props supplied globally to `<SnackbarsProvider>`, as the following demo illustrates:
```

samuelsycamore commented on docs/data/material/components/stacked-snackbars/stacked-snackbars.md:
```diff
@@ -0,0 +1,91 @@
+
+{{"demo": "PropsPriority.js"}}
+
+In the above example, the `anchorOrigin`, `autoHideDuration` and `TransitionComponent` props in `show` method are applied taking preference over the `props` supplied on `SnackbarsProvider`.
```
This can be cut for redundancy.

```suggestion
```

samuelsycamore commented on docs/data/material/components/stacked-snackbars/stacked-snackbars.md:
```diff
@@ -0,0 +1,91 @@
+
+## Global content and action props
+
+To have global consistency of the types of `action` and `content` every instance of snackbar has across your application, use the `ref` prop in the `SnackbarsProvider` component to apply the desired values to the aforementioned props.
```
Rephrasing to simplify the structure of clauses and pronouns.

```suggestion
You can add the `ref` prop to the `SnackbarsProvider` component to apply global values for the `action` and `content` props in all instances of snackbars across your entire application.
```

samuelsycamore commented on docs/data/material/components/stacked-snackbars/stacked-snackbars.md:
```diff
@@ -0,0 +1,91 @@
+
+To have global consistency of the types of `action` and `content` every instance of snackbar has across your application, use the `ref` prop in the `SnackbarsProvider` component to apply the desired values to the aforementioned props.
+
+This will save you from a lot of inefficient repetition if you'd to in each instance individually using the `show` method.
```
I think "defining default values" is more clear while also implying that it helps to prevent unnecessary repetition.

```suggestion
This lets you define default values rather than having to specify these props in every individual snackbar.
```

samuelsycamore commented on docs/data/material/components/stacked-snackbars/stacked-snackbars.md:
```diff
@@ -0,0 +1,91 @@
+To have global consistency of the types of `action` and `content` every instance of snackbar has across your application, use the `ref` prop in the `SnackbarsProvider` component to apply the desired values to the aforementioned props.
+
+This will save you from a lot of inefficient repetition if you'd to in each instance individually using the `show` method.
+
```
Add a line to introduce the demos.

```suggestion

The following demos show how to implement global values for the `action` prop and `content` prop, respectively:

```

samuelsycamore reviewed (commented):
Looking good! My review is focused primarily on the style, grammar, and clarity of the docs here. 👍

ZeeshanTamboli commented on docs/data/material/components/stacked-snackbars/stacked-snackbars.md:
```diff
@@ -0,0 +1,86 @@
+waiAria: https://www.w3.org/TR/wai-aria-1.1/#alert
+---
+
+# Stacked snackbars
```
Done by Danilo in https://github.com/mui/material-ui/pull/31991/commits/7a351f66c7efecb43be56c34a782349da7494594

ZeeshanTamboli commented:
@oliviertassinari 
> @ZeeshanTamboli I don't/won't have the time to dive into detail in this PR but for what I could see so far, we are on the right track 👍.

Great! I am excited to release this!

> Out of curiosity, I have found some of the leader in this space: https://npm-stat.com/charts.html?package=react-hot-toast&package=notistack&package=react-toastify&from=2021-05-22&to=2022-05-22. Likely projects we can learn from.

Sure.

> It seems that we miss the @mui/base story for this feature. I would be in favor we start there instead of with Material UI: so we build at the lower level first. The upvotes in [iamhosseindhv/notistack#184](https://github.com/iamhosseindhv/notistack/issues/184) suggest that there could be a large audience for this
> 
> <img alt="Screenshot 2022-05-23 at 01 48 33" width="100" src="https://user-images.githubusercontent.com/3165635/169721349-d31f7bc3-bacb-4c4d-91e3-135ea6b45f49.png">

Yeah, but now since the work is already almost done with the Material Design, I would like to keep it in Lab. Having it unstyled would be really useful. However I think devs can use the `content` prop to have their own custom components. Transition components (`Slide`, `Grow` etc) are the only MUI components this one is dependent on which can be removed in Base.

ZeeshanTamboli commented:
I was facing the problem of either having no new end of line space in `error-codes.json` file by doing `yarn extract-error-codes` but prettier wanted it to have a space. So I ignored this file in the prettier script in https://github.com/mui/material-ui/pull/31991/commits/81bed3bcca62c4d141d2cc35fe3f5de2f80071c8.

samuelsycamore commented:
Following a conversation with @mnajdova today, I think that the documentation for the stacked snackbars should live within the main Snackbar doc, following the format of the Tabs doc—see the [Experimental API section](https://mui.com/material-ui/react-tabs/#experimental-api). From a DX perspective, it makes more sense to keep the reader on a single page. If they need to build a UI with this functionality, they're going to check the Snackbars page before all else, and we don't want to make them jump around between pages.

oliviertassinari commented:
> Yeah, but now since the work is already almost done with the Material Design, I would like to keep it in Lab. Having it unstyled would be really useful. However most of the time, I think devs can use the content prop to have their own custom components. Transition components (Slide, Grow etc) are the only MUI components this one is dependent on which can be removed in Base.

To be clear, I was not question having the API in the lab, I think that this is great. My main concern is that it's a new feature, only for Material Design.

From a higher level consideration, we are moving toward @mui/material being entirely built on top of @mui/base in v6. So all features will first be available in MUI Base before Material UI or Joy UI. Now the question is, should we start now, as we have all the context of the logic in mind to structure the code to be compatible or do we ship to the lab, and then, somebody else will take on this task? Personally, I think that the former option is much better because it's a lot harder to course correct at a latter point in time. I would also +1 for making it a general rule, enforcing that any new features in Material UI should be only a styling layer on top of MUI Base so that we don't take step backwards on our target.

ZeeshanTamboli commented:
Thanks for the reviews. Summarizing the discussion here since there are multiple things and will help us what steps we should take next in this PR.

1. I have addressed all of the review comments except https://github.com/mui/material-ui/pull/31991#discussion_r850105670 and  https://github.com/mui/material-ui/pull/31991#issuecomment-1135612014.

2. Olivier suggesting to have it in Base as well along with Lab. I agree with his points seeing the upvotes in https://github.com/iamhosseindhv/notistack/issues/184#issue-505874617. I am just a bit hesitant doing it in this PR since `Snackbar` standalone component is not yet migrated (UnstyledSnackbar) as seen in https://github.com/mui/material-ui/issues/27170. I can work on it in a follow up PR. Or do you mean to have unstyled SnackbarsProvider or something along the lines of https://github.com/iamhosseindhv/notistack/issues/421#issuecomment-903198460? 

2. Move all the stacked snackbars demos and docs in Core's Snackbar component page, instead of having it in Lab. My point was that the implementation is done in `@mui/lab` so should we still show all the demos in Core? We should a stacked snackbar demo in the main Sanckbar page.

Let me know how I/we should proceed.

mnajdova commented:
> Thanks for the reviews. Summarizing the discussion here since there are multiple things and will help us what steps we should take next in this PR.
> 
> I have addressed all of the review comments except https://github.com/mui/material-ui/pull/31991#discussion_r850105670 and https://github.com/mui/material-ui/pull/31991#issuecomment-1135612014.

For `useId`. It is a hook, so if you want to use it outside of a component you cannot. But, you can get some inspiration from the `useId` implementation we had in v4 - https://github.com/mui/material-ui/blob/v4.x/packages/material-ui/src/utils/unstable_useId.js (before this hook existed in React).

On notistack I would say yes, if we have this API we should remove the demo. We can mention it as a inspiration maybe.

> Olivier suggesting to have it in Base as well along with Lab. I agree with his points seeing the upvotes in https://github.com/iamhosseindhv/notistack/issues/184#issue-505874617. I am just a bit hesitant doing it in this PR since Snackbar standalone component is not yet migrated (UnstyledSnackbar) as seen in https://github.com/mui/material-ui/issues/27170. I can work on it in a follow up PR. Or do you mean to have unstyled SnackbarsProvider or something along the lines of https://github.com/iamhosseindhv/notistack/issues/421#issuecomment-903198460?


I agree, having an unstyled `Snackbar` & `SnackbarProvider` would be great. Those can then be used in the lab for building the Material Design versions of them. I am fine with doing it in a follow up PR, or we can create a PR for adding the unstyled version, and put this one on hold till that one is done.

> Moving all the stacked snackbars demos and docs in Core's Snackbar component page, instead of having it in Lab. My point was that the implementation is done in @mui/lab so should we still show all the demos in Core? We show a stacked snackbar demo in the main Snackbar page.

I understand the direction, I would have done the same if I didn't know that we already are doing this for some other component :) Having all of them on the same page will also help with discoverability for this new API.

Let me know what you think @ZeeshanTamboli, but just wanted to mention that overall this has been a great execution.

ZeeshanTamboli commented:
Okay, I will proceed with the following steps in order:

1. - [ ] Keep this PR on hold and create `UnstyledSnackbar` component. It would also solve a part of Base #27170 . It will contain docs with demos as well in MUI Base.
2. - [ ] Merge this `UnstyledSnackbar` component in this PR and use it here for stacked snackbars in the lab implementation. Build it with Material design.
3. - [ ] Continuing in this PR, create `UnstyledSnackbarsProvider` and have the docs for it in MUI base. Basically a stacked snackbars page within MUI Base docs. (Unstyled stacked snackbars)
4. - [ ] Use `UnstyledSnackbarsProvider` within `SnackbarsProvider` in the lab implementation.
5. - [ ] Move the docs demos from MUI Lab to MUI core. Material design stacked Snackbars page will be removed from lab and we will have stacked snackbars demos within Core `Snackbar` component page.

@mnajdova What do you think? 

cc: @oliviertassinari @danilo-leal @samuelsycamore

mnajdova commented:
> @mnajdova What do you think?

Sounds great 👍

Nic-S commented:
I read with pleasure about this new feature, if I may I would have a consideration. The main advantage of [react-tostify](https://fkhadra.github.io/react-toastify/introduction) is the fact that you can use it outside the lifecycle of a react component, two practical examples of this use might be: 
- invoke the snackbar within an action if the application uses redux-toolkit or other state management (in this case perhaps you can mitigate by passing the result of the useSnackbars hook but would need to find a way not to pass it every time you invoke an action that requires a snackbar)
- invoke the snackbar inside an interceptor, for example an axios interceptor that handles some generic errors.
 @ZeeshanTamboli Is there a way to expose the show() and close() api for the above examples?

ZeeshanTamboli commented:
> I read with pleasure about this new feature, if I may I would have a consideration. The main advantage of [react-tostify](https://fkhadra.github.io/react-toastify/introduction) is the fact that you can use it outside the lifecycle of a react component, two practical examples of this use might be:
> 
> * invoke the snackbar within an action if the application uses redux-toolkit or other state management (in this case perhaps you can mitigate by passing the result of the useSnackbars hook but would need to find a way not to pass it every time you invoke an action that requires a snackbar)
> * invoke the snackbar inside an interceptor, for example an axios interceptor that handles some generic errors.


Thanks @Nic-S for taking a look!

> @ZeeshanTamboli Is there a way to expose the show() and close() api for the above examples?

I am not quite understanding your examples without a code example. Anyway, we want to first release this `useSnackbars` hook and then we can iterate more on it and add new features.

Hummel37 commented:
Would it be possible to display a notification wihtout using the `useSnackbars()` hook?
Like the `toast `method in https://github.com/fkhadra/react-toastify.
That would really be nice 😄

FaabLondon commented:
Hello,
thanks for this https://deploy-preview-31991--material-ui.netlify.app/material-ui/react-stacked-snackbars/#api
in which version of mui/lab is it available? It does not seem to be available in 5.0.0-alpha.112
thanks

jpsala commented:
> Hello, thanks for this https://deploy-preview-31991--material-ui.netlify.app/material-ui/react-stacked-snackbars/#api in which version of mui/lab is it available? It does not seem to be available in 5.0.0-alpha.112 thanks

It's not there yet it seems :(

jpsala commented:
Hi, there is a release date for this feature?

ZeeshanTamboli commented:
> Hi, there is a release date for this feature?

There is no release date. We are planning to build unstyled stacked snackbars first. And then we will rebuild this on top of that unstyled version.

jpsala commented:
> > Hi, there is a release date for this feature?
> 
> There is no release date. We are planning to build unstyled stacked snackbars first. And then we will rebuild this on top of that unstyled version.

Thank you!

ZeeshanTamboli commented:
@mui/core I moved the [docs](https://deploy-preview-31991--material-ui.netlify.app/material-ui/react-snackbar/#experimental-api) from Lab to Material UI Snackbar page.

gnowland commented:
> > Hi, there is a release date for this feature?
> 
> There is no release date. We are planning to build unstyled stacked snackbars first. And then we will rebuild this on top of that unstyled version.

Current progress: https://github.com/mui/material-ui/pull/31991#issuecomment-1148374025

(Since the "progress" comment is now hidden under "Load more..." (thanks, github 🙄 -- upvote the "pinned comment" feature request: https://github.com/orgs/community/discussions/47912))

gnowland commented:
> > Hi, there is a release date for this feature?
> 
> There is no release date. We are planning to build unstyled stacked snackbars first. And then we will rebuild this on top of that unstyled version.

@ZeeshanTamboli do we have to wait until _all_ unstyled components (https://github.com/mui/material-ui/issues/27170) are completed, or now that the unstyled snackbar component is complete is work on checkbox 2 able to proceed and this component be available independently of _all_ unstyled components being completed?

aarongarciah commented:
Closing due to inactivity. Feel free to reopen.
